File: sources/Sleek v0.1/me/kansio/client/modules/impl/combat/TargetStrafe.java
package me.kansio.client.modules.impl.combat;

import com.google.common.eventbus.Subscribe;
import me.kansio.client.Client;
import me.kansio.client.event.impl.Render3DEvent;
import me.kansio.client.event.impl.UpdateEvent;
import me.kansio.client.modules.api.ModuleCategory;
import me.kansio.client.modules.api.ModuleData;
import me.kansio.client.modules.impl.Module;
import me.kansio.client.value.value.BooleanValue;
import me.kansio.client.value.value.NumberValue;
import net.minecraft.entity.Entity;

import static org.lwjgl.opengl.GL11.*;

@ModuleData(
        name = "Target Strafe",
        category = ModuleCategory.COMBAT,
        description = "Automatically strafes around the killaura target"
)
public class TargetStrafe extends Module {

    public static double dir = -1;

    public BooleanValue autoF5 = new BooleanValue("Auto F5", this, false);
    public BooleanValue jump = new BooleanValue("On Jump", this, false);
    public BooleanValue control = new BooleanValue("Controllable", this, false);
    public BooleanValue render = new BooleanValue("Render", this, false);
    public NumberValue<Double> radius = new NumberValue<>("Radius", this, 3d, 0.1d, 5d, 0.1);
    public NumberValue<Double> width = new NumberValue<>("Width", this, 1d, 0.1d, 5d, 0.1, render);
    public NumberValue<Double> red = new NumberValue<>("Red", this, 100d, 0d, 255d, 1d, render);
    public NumberValue<Double> green = new NumberValue<>("Green", this, 100d, 0d, 255d, 1d, render);
    public NumberValue<Double> blue = new NumberValue<>("Blue", this, 100d, 0d, 255d, 1d, render);

    @Override
    public void onDisable() {
        dir = -1;
    }

    @Subscribe
    public void onMotion(UpdateEvent event) {
        if (canStrafe() && autoF5.getValue()) {
            mc.gameSettings.thirdPersonView = 1;
        } else if (!canStrafe() && autoF5.getValue()) {
            mc.gameSettings.thirdPersonView = 0;
        }

        if (event.isPre()) {
            if (control.getValue()) {
                if (mc.gameSettings.keyBindLeft.isPressed()) {
                    dir = 1;
                } else if (mc.gameSettings.keyBindRight.isPressed()) {
                    dir = -1;
                }
            }

            if (mc.thePlayer.isCollidedHorizontally) {
                invertStrafe();
            }
        }
    }

    private void invertStrafe() {
        dir = -dir;
    }

    @Subscribe
    public void onRender(Render3DEvent event) {
        if (canStrafe() && render.getValue()) {
            drawCircle(KillAura.target, mc.timer.renderPartialTicks);
        }
    }

    public boolean canStrafe() {
        if (!Client.getInstance().getModuleManager().getModuleByName("KillAura").isToggled()) {
            return false;
        }

        if (!this.isToggled()) {
            return false;
        }

        if (jump.getValue()) {
            if (mc.gameSettings.keyBindJump.isKeyDown())
                return KillAura.target != null;
            else
                return false;
        } else {
            return KillAura.target != null;
        }
    }

    private void drawCircle(Entity entity, float partialTicks) {
        glPushMatrix();
        glColor3d(red.getValue(), green.getValue(), blue.getValue());
        glDisable(GL_TEXTURE_2D);
        glDisable(GL_DEPTH_TEST);
        glDepthMask(false);
        glLineWidth(width.getValue().floatValue());
        glBegin(GL_LINE_STRIP);

        final double x = entity.lastTickPosX + (entity.posX - entity.lastTickPosX) * partialTicks - mc.getRenderManager().viewerPosX;
        final double y = entity.lastTickPosY + (entity.posY - entity.lastTickPosY) * partialTicks - mc.getRenderManager().viewerPosY;
        final double z = entity.lastTickPosZ + (entity.posZ - entity.lastTickPosZ) * partialTicks - mc.getRenderManager().viewerPosZ;


        final double pix2 = Math.PI * 2.0D;
        for (int i = 0; i <= 90; ++i) {
            glVertex3d(x + (radius.getValue() - 0.5) * Math.cos(i * pix2 / 45), y, z + (radius.getValue() - 0.5) * Math.sin(i * pix2 / 45));
        }

        glEnd();
        glDepthMask(true);
        glEnable(GL_DEPTH_TEST);
        glEnable(GL_TEXTURE_2D);
        glPopMatrix();
    }
}

File: sources/Kevin Reborn/kevin/module/modules/movement/TargetStrafe.kt
/*
 * This program is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation, either version 3 of the License, or
 *   (at your option) any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package kevin.module.modules.movement

import kevin.event.*
import kevin.main.KevinClient
import kevin.module.*
import kevin.utils.MovementUtils
import kevin.utils.RotationUtils
import net.minecraft.entity.EntityLivingBase
import org.lwjgl.opengl.GL11
import java.awt.Color
import kotlin.math.cos
import kotlin.math.sin

object TargetStrafe : Module("TargetStrafe","Strafe around your target.", category = ModuleCategory.MOVEMENT) {
    private val radius = FloatValue("Radius", 2.0f, 0.1f, 4.0f)
    private val strengthValue= FloatValue("Strength", 0.5F, 0F, 1F)
    private val render = BooleanValue("Render", true)
    private val space = BooleanValue("HoldSpace", false)
    private val hurtTime = IntegerValue("MinHurtTime", 0, 0, 9)
    private val safewalk = BooleanValue("SafeWalk", true)
    private val onlySpeed = BooleanValue("OnlySpeed", false)
    private var direction = -1

    @EventTarget
    fun onMotion(event: MotionEvent) {
        if (event.eventState === EventState.PRE) {
            if (mc.thePlayer.isCollidedHorizontally) {
                switchDirection()
            }
            if (mc.gameSettings.keyBindLeft.isKeyDown) {
                direction = 1
            }
            if (mc.gameSettings.keyBindRight.isKeyDown) {
                direction = -1
            }
        }
    }

    private fun switchDirection() {
        direction = if (direction == 1) -1 else 1
    }

    @EventTarget
    fun onMove(event: MoveEvent) {
        if (safewalk.get() && mc.thePlayer.onGround && canStrafe(KevinClient.combatManager.target))
            event.isSafeWalk = true

        val target = KevinClient.combatManager.target
        if (canStrafe(target))
            setSpeed(event, MovementUtils.speed.toDouble(), RotationUtils.getRotationsEntity(target).yaw, direction.toDouble(), if (mc.thePlayer.getDistanceToEntity(target) <= radius.get()) 0.0 else 1.0)
    }

    private fun setSpeed(moveEvent: MoveEvent, moveSpeed: Double, pseudoYaw: Float, pseudoStrafe: Double, pseudoForward: Double) {
        var forward = pseudoForward
        var strafe = pseudoStrafe
        var yaw = pseudoYaw

        if (forward == .0 && strafe == .0) {
            moveEvent.z = .0
            moveEvent.x = .0
        } else {
            if (forward != .0) {
                if (strafe > .0) {
                    yaw += (if (forward > .0) -45 else 45)
                } else if (strafe < .0) {
                    yaw += (if (forward > .0) 45 else -45)
                }
                strafe = 0.0
                if (forward > .0) {
                    forward = 1.0
                } else if (forward < .0) {
                    forward = -1.0
                }
            }
            val cos = cos(Math.toRadians(yaw + 90.0))
            val sin = sin(Math.toRadians(yaw + 90.0))

            val speed = moveSpeed * strengthValue.get()
            val motionX = (mc.thePlayer!!.motionX * (1 - strengthValue.get()))
            val motionZ = (mc.thePlayer!!.motionZ * (1 - strengthValue.get()))

            moveEvent.x = (forward * speed * cos + strafe * speed * sin) + motionX
            moveEvent.z = (forward * speed * sin - strafe * speed * cos) + motionZ
        }
    }

    @EventTarget
    fun onRender3D(event: Render3DEvent) {
        val target = KevinClient.combatManager.target
        if (canStrafe(target) && render.get()) {
            target?:return
            GL11.glPushMatrix()
            GL11.glDisable(3553)
            GL11.glEnable(2848)
            GL11.glEnable(2881)
            GL11.glEnable(2832)
            GL11.glEnable(3042)
            GL11.glBlendFunc(770, 771)
            GL11.glHint(3154, 4354)
            GL11.glHint(3155, 4354)
            GL11.glHint(3153, 4354)
            GL11.glDisable(2929)
            GL11.glDepthMask(false)
            GL11.glLineWidth(1.0f)

            GL11.glBegin(3)
            val x = target.lastTickPosX + (target.posX - target.lastTickPosX) * event.partialTicks - mc.renderManager.viewerPosX
            val y = target.lastTickPosY + (target.posY - target.lastTickPosY) * event.partialTicks - mc.renderManager.viewerPosY
            val z = target.lastTickPosZ + (target.posZ - target.lastTickPosZ) * event.partialTicks - mc.renderManager.viewerPosZ
            for (i in 0..360) {
                val rainbow = Color(Color.HSBtoRGB((mc.thePlayer.ticksExisted / 70.0 + sin(i / 50.0 * 1.75)).toFloat() % 1.0f, 0.7f, 1.0f))
                GL11.glColor3f(rainbow.red / 255.0f, rainbow.green / 255.0f, rainbow.blue / 255.0f)
                GL11.glVertex3d(x + radius.get() * cos(i * 6.283185307179586 / 45.0), y, z + radius.get() * sin(i * 6.283185307179586 / 45.0))
            }
            GL11.glEnd()

            GL11.glDepthMask(true)
            GL11.glEnable(2929)
            GL11.glDisable(2848)
            GL11.glDisable(2881)
            GL11.glEnable(2832)
            GL11.glEnable(3553)
            GL11.glPopMatrix()
        }
    }

    private fun canStrafe(target: EntityLivingBase?): Boolean {
        return this.state && target != null && (!space.get() || mc.thePlayer.movementInput.jump) && (!onlySpeed.get() || KevinClient.moduleManager.getModule(Speed::class.java).state) && (hurtTime.get() <= mc.thePlayer.hurtTime)
    }
}